Defender behaviour (#1127)
NUbots/NUbots: module/purpose/Defender/README.md

# Defender

## Description

Play soccer in the defender position, which is an defensive position. Either bypass GameController and force play, or uses GameController information to decide what to do.

> In the ready state, makes a Ready subtask.

Reviewer: ```suggestion
In the ready state, walks to a position on the field.
```

# Defender

## Description

Play soccer in the defender position, which is an defensive position. Either bypass GameController and force play, or uses GameController information to decide what to do.

In the ready state, makes a Ready subtask.

> In the playing state (for both normal game mode and penalty shootout), finds the ball, fixates on the ball, walks to the ball if the ball is in the defenders bounded box defensive area, and kicks the ball.

Reviewer: we don't care about penalty shootout, it's not applicable for the defender

If ball is outside bounded box area and ball is in opposition half, defender is clamped to the edge of bounding box closest to ball.

If ball is outside bounded box area and ball is in own half, defender is clamped to the edge of bounding box closest to ball and position itself 1 meter behind the ball.

## Usage

Add this module to play as a defender in soccer!

## Consumes

- `message::input::GameState` to get information about the state of the game
- `message::input::GameState::Data::Phase` to get specific information about the current game phase (initial, ready, set, playing, etc).

> - `message::strategy::WalkInsideBounded` a Task requesting to play as a Defender

Reviewer: ```suggestion
- `message::purpose::Defender` a Task requesting to play as a Defender
```

Reviewer: you didn't apply the suggestion above. WalkInsideBounded is not consumed by this module, but Defender is.

## Emits

> - `message::strategy::WalkInsideBoundedBox` a Task requesting to play as a defender in the normal game state clamped to bounded box

Reviewer: this doesn't make sense to me, the task is just saying to keep within a bounding box. the task doesn't say to play as a defender, the Defender task does that

Reviewer: ```suggestion
- `message::strategy::WalkInsideBoundedBox` a Task requesting to stay within a bounded area on the field.
```

> - `message::strategy::Ready` a Task requesting to walk to the ready position

Reviewer: do you use this? I'm pretty sure this is an old module that we don't use

> # Striker

Reviewer: Copy paste of Striker?

- `message::input::GameState` to get information about the state of the game
- `message::input::GameState::Data::Phase` to get specific information about the current game phase (initial, ready, set, playing, etc).

> - `message::strategy::Denfender` a Task requesting to play as a Defender

Reviewer: ```suggestion
- `message::strategy::Defender` a Task requesting to play as a Defender
```
